pkg/issuer/acme/dns/digitalocean/digitalocean.go

```diff
@@ -0,0 +1,190 @@
+
+	// if DigitalOcean does not have this zone then we will find out later
+	zoneName, err := util.FindZoneByFqdn(fqdn, util.RecursiveNameservers)
+	if err != nil {
+		return err
+	}
+
+	rec := digitalOceanRecord{
+		Type: "TXT",
+		Name: util.UnFqdn(fqdn),
+		Data: value,
+		TTL:  120,
+	}
+
+	body, err := json.Marshal(rec)
+	if err != nil {
+		return err
+	}
+
+	_, err = c.makeRequest("POST", fmt.Sprintf("/domains/%s/records", util.UnFqdn(zoneName)), bytes.NewReader(body))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// CleanUp removes the TXT record matching the specified parameters
+func (c *DNSProvider) CleanUp(domain, token, keyAuth string) error {
+	fqdn, _, _ := util.DNS01Record(domain, keyAuth)
+
+	zoneName, err := util.FindZoneByFqdn(fqdn, util.RecursiveNameservers)
+
+	record, err := c.findTxtRecord(fqdn)
+	if err != nil {
+		return err
+	}
+
+	_, err = c.makeRequest("DELETE", fmt.Sprintf("/domains/%s/records/%d", util.UnFqdn(zoneName), record.ID), nil)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *DNSProvider) findTxtRecord(fqdn string) (*digitalOceanRecord, error) {
+
+	// APIDomainRecordsResponse represents a domain records list response from the digitalocean API
+	type APIDomainRecordsResponse struct {
+		DomainRecords []digitalOceanRecord `json:"domain_records"`
+	}
+
+	zoneName, err := util.FindZoneByFqdn(fqdn, util.RecursiveNameservers)
+	if err != nil {
+		return nil, err
+	}
+
+	result, err := c.makeRequest(
+		"GET",
+		fmt.Sprintf("/domains/%s/records?per_page=1000&type=TXT&name=%s", util.UnFqdn(zoneName), util.UnFqdn(fqdn)),
+		nil,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	var records APIDomainRecordsResponse
+	err = json.Unmarshal(result, &records)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, rec := range records.DomainRecords {
+		if rec.Name == util.UnFqdn(fqdn) {
+			return &rec, nil
+		}
+	}
+
+	return nil, fmt.Errorf("No existing record found for %s", fqdn)
+}
+
+func (c *DNSProvider) makeRequest(method, uri string, body io.Reader) (json.RawMessage, error) {
+	// APIError contains error details for failed requests
+	type APIError struct {
+		Code    int    `json:"code,omitempty"`
+		Message string `json:"message,omitempty"`
+	}
+
+	req, err := http.NewRequest(method, fmt.Sprintf("%s%s", DigitalOceanAPIURL, uri), body)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.authToken))
+	req.Header.Set("Content-Type", "application/json")
+	//req.Header.Set("User-Agent", userAgent())
+
+	client := http.Client{Timeout: 30 * time.Second}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("Error querying DigitalOcean API -> %v", err)
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 400 {
+		var r APIError
+
+		err = json.NewDecoder(resp.Body).Decode(&r)
+		if err != nil {
+			return nil, err
+		}
+
+		return nil, fmt.Errorf("DigitalOcean API Error \n%d: %s", r.Code, r.Message)
```

Comment:
In general, it's preferable to not have newlines in error strings.

Reply:
I don't know go very well, I was just copying from the other DNS providers. Will fix if @munnerz agrees.

Reply:
yep could you remove the `\n` 😄